### PR TITLE
3161 remove i control channel

### LINKED ISSUE #3161: Remove IControlChannel
The ControlChannel currently serves as a facade for the `AutomatedMaster`, `InfectionMonkey` and `AggregatingPropagationCredentialsRepository`. However, it has become less and less useful over time and can now be completely removed.

Suggestion:
- Remove `IControlChannel.get_config`, and replace it with `IIslandAPIClient.get_config`
- Remove `IControlChannel.get_credentials_for_propagation`. `AggregatingPropagationCredentialsRepository` is the sole user. We can use the one provided by `IIslandAPIClient` instead.
- Replace `IControlChannel.should_agent_stop`. This could likely be its own component

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes #3161.

Add any further explanations here.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [x] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/infection_monkey/propagation_credentials_repository/aggregating_propagation_credentials_repository.py
Comment (shreyamalviya) [inline monkey/infection_monkey/propagation_credentials_repository/aggregating_propagation_credentials_repository.py:1]: Why did you add this back?

File: monkey/infection_monkey/utils/agent_process.py [hunk 1]
```diff
@@ -1,8 +1,22 @@
...
     agent_process = psutil.Process(os.getpid())
     return agent_process.create_time()
+
+
+def should_agent_stop(server: Optional[str], island_api_client: IIslandAPIClient) -> bool:
+    if not server:
+        logger.error("Agent should stop because it can't connect to the C&C server.")
+        return True
+    agent_signals = island_api_client.get_agent_signals()
+    return agent_signals.terminate is not None
```
Comment (shreyamalviya) [inline monkey/infection_monkey/utils/agent_process.py:22]: It doesn't make sense to put this function in this file.  `IIslandAPIClient` should be hiding all the Island-related logic from the rest of the Agent.

File: monkey/infection_monkey/master/automated_master.py [hunk 2]
```diff
@@ -26,16 +27,18 @@
 class AutomatedMaster(IMaster):
     def __init__(
         self,
+        island_address: str,
         current_depth: Optional[int],
         servers: Sequence[str],
         puppet: IPuppet,
-        control_channel: IControlChannel,
+        island_api_client: IIslandAPIClient,
```
Comment (shreyamalviya) [inline monkey/infection_monkey/master/automated_master.py:34]: The whole point of having `IIslandAPIClient` is so that we don't leak any Island details into the Agent. `island_address` shouldn't be passed here.

**Review by shreyamalviya — CHANGES_REQUESTED** [review]
Move `should_agent_stop()` into `IIslandAPIClient` and use it through the client.

File: monkey/infection_monkey/master/agent_lifecycle.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from infection_monkey.island_api_client import IIslandAPIClient
+
+
+class AgentLifecycle:
+    def __init__(self, island_api_client: IIslandAPIClient):
+        self._island_api_client = island_api_client
+
+    def should_agent_stop(self) -> bool:
+        agent_signals = self._island_api_client.get_agent_signals()
+        return agent_signals.terminate is not None
```
Comment (mssalvatore) [inline monkey/infection_monkey/master/agent_lifecycle.py:10]: I don't think this component makes any sense. We've taken a thin component with very little functionality (IControlClient) and replaced it with an even thinner component with less functionality (AgentLifecycle). The two lines in this function can just be inlined into `AutomatedMaster._check_for_stop()`.

Alternatively, we could add `IIslandAPIClient.terminate_signal_set()` as a convenience function.

File: monkey/infection_monkey/monkey.py [hunk 4]
```diff
@@ -519,10 +519,10 @@ def _stop_relay(self):
         if self._relay and self._relay.is_alive():
             self._relay.stop()
 
-            while self._relay.is_alive() and not self._control_channel.should_agent_stop():
+            while self._relay.is_alive() and not self._agent_lifecycle.should_agent_stop():
                 self._relay.join(timeout=5)
 
-            if self._control_channel.should_agent_stop():
+            if self._agent_lifecycle.should_agent_stop():
                 self._relay.join(timeout=60)
```
Comment (mssalvatore) [inline monkey/infection_monkey/monkey.py:526]: I've spotted a bug. It's possible that the agent is shutting down because communication with the Island has been disrupted. This means `should_agent_shop()` could raise an exception, causing a number of cleanup operations in `cleanup()` to be skipped.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/monkey.py (hunk 4 above)
Comment (mssalvatore) [reply monkey/infection_monkey/monkey.py]: @ilija-lazoroski Fixed. Take a look. We should probably run and observe tunneling tests, but I think the change are fairly straight forward.

**Review by mssalvatore — APPROVED** [review]
